src/audio/codec_adapter/codec/generic.c

```diff
@@ -346,30 +280,119 @@ void module_free_all_memory(struct comp_dev *dev)
 	}
 }
 
-int module_free(struct comp_dev *dev)
+int module_free(struct processing_module *mod)
 {
 	int ret;
-	struct processing_module *mod = comp_get_drvdata(dev);
 	struct module_data *md = &mod->priv;
 
-	ret = md->ops->free(dev);
+	ret = md->ops->free(mod);
 	if (ret)
-		comp_warn(dev, "module_free(): error: %d for module_id %x",
-			  ret, mod->ca_config.module_id);
+		comp_warn(mod->dev, "module_free(): error: %d for %d",
+			  ret, dev_comp_id(mod->dev));
 
 	/* Free all memory requested by module */
-	module_free_all_memory(dev);
+	module_free_all_memory(mod);
 	/* Free all memory shared by codec_adapter & module */
-	md->s_cfg.avail = false;
-	md->s_cfg.size = 0;
-	md->r_cfg.avail = false;
-	md->r_cfg.size = 0;
-	rfree(md->r_cfg.data);
-	rfree(md->s_cfg.data);
+	md->cfg.avail = false;
+	md->cfg.size = 0;
+	rfree(md->cfg.data);
 	if (md->runtime_params)
 		rfree(md->runtime_params);
 
 	md->state = MODULE_DISABLED;
 
 	return ret;
 }
+
+/**
+ * \brief Set module configuration - Common method to assemble large configuration message
+ * \param[in] mod - struct processing_module pointer
+ * \param[in] config_id - Configuration ID
+ * \param[in] pos - position of the fragment in the large message
+ * \param[in] data_offset_size: size of the whole configuration if it is the first fragment or the
+ *				 only fragment. Otherwise, it is the offset of the fragment in the
+ *				 whole configuration.
+ * \param[in] fragment: configuration fragment buffer
+ * \param[in] fragment_size: size of @fragment
+ * \params[in] response: optional response buffer to fill
+ * \params[in] response_size: size of @response
+ *
+ * \return: 0 upon success or error upon failure
+ */
+int module_set_configuration(struct processing_module *mod,
+			     uint32_t config_id,
+			     enum module_cfg_fragment_position pos, uint32_t data_offset_size,
+			     const uint8_t *fragment, size_t fragment_size, uint8_t *response,
+			     size_t response_size)
+{
+	struct module_data *md = &mod->priv;
+	static uint32_t size;
+	uint32_t offset = 0;
+	char *dst;
+	int ret;
```

Comment:
@ranj063 I get a compiler warning in this function as `ret` is used in a trace message before it's assigned

Reply:
Thanks for reporting this @markbartonxperi . I am still working to fix the issues with the cadence codecs with #5190 . I will update this PR after that one is merged